feat(tempdeck-gen3): add queue aggregator

fsinapi opened the pull request:
This PR adds a "queue aggregator" type, intended to replace the hard-coded central task list for TD3 firmware. The class maintains an internal tuple of queue handles (which are just wrappers around pointers to the queue types), and provides a few different ways to send a message to a specific queue. 

- Tests for QueueAggregator are added
- Consolidated TestMessageQueue to be in the `common` include folder instead of being individually defined for Heater-Shaker and Thermocycler-Gen2
- Modified the TestMessageQueue type in order to fit with the requirements of the QueueAggregator
- The QueueAggregator is not used in anything outside of tests so far

fsinapi committed:
WIP version of queue aggregator + tests

fsinapi committed:
Got index-based runtime lookups to work

fsinapi committed:
Added some better documentation on queue aggregator, and cleaned up tests

fsinapi committed:
Slightly more in-depth testing

fsinapi committed:
Added a constructor to the QueueAggregator

fsinapi committed:
Added changes to TestMessageQueue to enable cloned message queue types

fsinapi committed:
Updated some comments

fsinapi committed:
Re-added a commented out assertion

sfoster1 commented on stm32-modules/include/common/core/queue_aggregator.hpp:
```diff
@@ -0,0 +1,274 @@
+     * @return size_t A unique index for this message type
+     */
+    template <typename Queue>
+    constexpr static auto get_task_idx() -> size_t {
```
can we use `consteval` here, which will cause a compilation failure if this can't be evaluated during compilation

sfoster1 commented on stm32-modules/include/common/core/queue_aggregator.hpp:
```diff
@@ -0,0 +1,274 @@
+     * to forward it to.
+     *
+     * @tparam Tag The tag type of the queue to send to. This must be
+     * provided by each queue type delcared for the QueueAggregator.
```
```suggestion
     * provided by each queue type declared for the QueueAggregator.
```

sfoster1 commented on stm32-modules/include/common/core/queue_aggregator.hpp:
```diff
@@ -0,0 +1,274 @@
+    template <typename Tag, typename Message>
+    auto send(const Message& msg, const Tag& tag = (Tag())) -> bool {
+        static_cast<void>(tag);
+        constexpr auto idx = get_task_idx<Tag>();
```
Isn't `get_task_idx`'s tparam the `Queue` and it derives the tag from there? How does this call work?

sfoster1 reviewed (changes requested):
The functionality looks good but I'm a bit confused by how `get_task_idx` works when you give it a `Tag`-type as a template param instead of a `Queue` type. Either there's a bug in the code that somehow doesn't prevent compilation or we really gotta change some of the names and the docs to match the functionality.

fsinapi committed:
Update stm32-modules/include/common/core/queue_aggregator.hpp

Co-authored-by: Seth Foster <seth@opentrons.com>

fsinapi committed:
Renamed `get_task_idx` to `get_queue_idx` for disambiguation
- Added dedicated `get_tag_idx` for code readability
- Cleaned up some innacurate function comments
- Use `consteval` where appropriate

fsinapi committed:
Added a concept constraint for `get_queue_idx`

sfoster1 reviewed (approved):
Awesome! Can't wait to see this in practice.
